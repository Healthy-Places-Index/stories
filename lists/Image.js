const { Text, Url, Relationship } = require('@keystonejs/fields');

const defaultAuth = ({ existingItem, authentication: { item } }) =>
  item && existingItem.slide.project.user.toString() === item.id;

module.exports = {
  access: {
    auth: true,
    create: ({ authentication: { item } }) => item !== undefined,
    read: true,
    update: defaultAuth,
    delete: defaultAuth,
  },
  fields: {
    title: {
      type: Text,
    },
    creator: {
      type: Text,
    },
    source: {
      type: Text,
    },
    date: {
      type: Text,
    },
    url: {
      type: Url,
    },
    slide: {
      type: Relationship,
      ref: 'Slide.image',
      many: false,
    },
  },
  labelField: 'title',
};
