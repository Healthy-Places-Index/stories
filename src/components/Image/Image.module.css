.closeButton{
  position: absolute;
  top: 5px;
  right: 5px;
}

.inlineLabel{
  width: 45px !important;
}

.inlineInput{
  width: calc(100% - 57px) !important;
}