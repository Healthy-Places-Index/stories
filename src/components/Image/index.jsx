/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import { Segment, Image as Img, Form, Input, Button } from 'semantic-ui-react';

import styles from './Image.module.css';

const Image = ({ image, updateHandler }) => {
  const fileInputRef = useRef(null);

  const [isLoading, setIsLoading] = useState(false);
  const [urlError, setUrlError] = useState(false);
  const [imageMeta, setImageMeta] = useState(null);

  useEffect(() => {
    if (image) {
      setImageMeta({
        imageTitle: image.imageTitle,
        source: image.source,
      });
    }
  }, [image]);

  const getSignedUrl = e => {
    const file = e.target.files[0];
    if (file) {
      setIsLoading(true);
      return axios.post('/upload', { name: file.name, type: file.type }).then(({ data }) => {
        return axios
          .put(data.signedRequest, file, { headers: { 'Content-Type': file.type } })
          .then(() => {
            setIsLoading(false);
            updateHandler(image.id, { url: data.url }, 1);
          });
      });
    }
    return null;
  };

  return (
    <Segment>
      {image?.url ? (
        <>
          <Img src={image.url} />
          <Button
            className={styles.closeButton}
            circular
            icon="close"
            size="mini"
            onClick={() => updateHandler(image.id, { url: null }, 1)}
          />
        </>
      ) : (
        <>
          <Button
            content="Choose File"
            labelPosition="left"
            icon="file"
            onClick={() => fileInputRef.current.click()}
            loading={isLoading}
          />
          <input ref={fileInputRef} type="file" hidden onChange={getSignedUrl} />
        </>
      )}
      <Form.Field inline style={{ marginTop: 15 }} disabled={!image?.url}>
        <label className={styles.inlineLabel}>Caption</label>
        <Input
          placeholder="Image caption"
          className={styles.inlineInput}
          value={imageMeta && image.url ? imageMeta.imageTitle || '' : ''}
          onChange={(e, { value }) => {
            setImageMeta({ ...imageMeta, imageTitle: value });
            updateHandler(image.id, { imageTitle: value });
          }}
        />
      </Form.Field>
      <Form.Field inline disabled={!image?.url} error={urlError}>
        <label className={styles.inlineLabel}>Link</label>
        <Input
          placeholder="https://..."
          className={styles.inlineInput}
          value={imageMeta ? imageMeta.source || '' : ''}
          onChange={(e, { value }) => {
            setImageMeta({ ...imageMeta, source: value });
            if (
              !value ||
              value.match(
                /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_+.~#?&//=]*)/
              )
            ) {
              updateHandler(image.id, { source: value });
              setUrlError(false);
            } else {
              setUrlError(true);
            }
          }}
        />
      </Form.Field>
    </Segment>
  );
};

Image.propTypes = {
  image: PropTypes.shape(),
  updateHandler: PropTypes.func.isRequired,
};

Image.defaultProps = {
  image: null,
};

export default Image;
