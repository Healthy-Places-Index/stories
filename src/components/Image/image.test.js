import React from 'react';
import renderer, { act } from 'react-test-renderer';

import Image from './index';

jest.mock('semantic-ui-react', () => {
  const Button = props => <div {...props} className="button" />;
  const Segment = props => <div {...props} className="segment" />;
  const Field = props => <div {...props} className="field" />;
  const Form = { Field };
  const Input = props => <input {...props} className="input" />;
  return { Button, Segment, Form, Input };
});

describe('Image', () => {
  it('matches snapshot', async () => {
    const component = renderer.create(
      <Image
        image={{ imageTitle: 'Test image', creator: 'Test creator', source: 'Test source' }}
        updateHandler={() => {}}
      />
    );

    await act(() => new Promise(resolve => setTimeout(resolve, 0)));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
