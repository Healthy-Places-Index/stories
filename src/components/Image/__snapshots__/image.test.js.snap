// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image matches snapshot 1`] = `
<div
  className="segment"
>
  <div
    className="button"
    content="Choose File"
    icon="file"
    labelPosition="left"
    loading={false}
    onClick={[Function]}
  />
  <input
    hidden={true}
    onChange={[Function]}
    type="file"
  />
  <div
    className="field"
    disabled={true}
    inline={true}
    style={
      Object {
        "marginTop": 15,
      }
    }
  >
    <label
      className="inlineLabel"
    >
      Title
    </label>
    <input
      className="input"
      onChange={[Function]}
      value="Test image"
    />
  </div>
  <div
    className="field"
    disabled={true}
    inline={true}
  >
    <label
      className="inlineLabel"
    >
      Creator
    </label>
    <input
      className="input"
      onChange={[Function]}
      value="Test creator"
    />
  </div>
  <div
    className="field"
    disabled={true}
    inline={true}
  >
    <label
      className="inlineLabel"
    >
      Source
    </label>
    <input
      className="input"
      onChange={[Function]}
      value="Test source"
    />
  </div>
</div>
`;
